### PR TITLE
Remove functionality

### PR DESCRIPTION
**Pull request by mcdonnelldean**
@mcollina Can you review please? Particularly for any side effects to the filters?

I also went back to the original test style since it matches with all of your other repositories.

### CODE REVIEW

File: bloomrun.js [hunk 1]
```diff
@@ -43,6 +43,40 @@ BloomRun.prototype.add = function (pattern, payload) {
...
 }
 
+BloomRun.prototype.remove = function (pattern) {
+  var properties = this._properties
+  var matches = matchingBuckets(this._buckets, pattern)
+  var buckets = this._buckets
+  var bucket
+
+  if (matches.length > 0) {
+    bucket = matches[0]
```
Comment (mcollina) [inline bloomrun.js]: You need to iterate here, if the first bucket do not contain the pattern.

File: bloomrun.js [hunk 2]
```diff
@@ -43,6 +43,40 @@ BloomRun.prototype.add = function (pattern, payload) {
...
+
+    for (var i = 0; i < bucket.data.length; i++) {
+      if (pattern === bucket.data[i].pattern) {
+        bucket.data.splice(i, 1)
+      }
```
Comment (mcollina) [inline bloomrun.js]: You should also check about the second argument off `add`.

File: bloomrun.js [hunk 3]
```diff
@@ -43,6 +43,40 @@ BloomRun.prototype.add = function (pattern, payload) {
...
+    }
+
+    for (var b = 0; i < buckets.length; i++) {
+      if (bucket === buckets[b]) {
+        this._buckets.splice(b, 1)
+      }
+    }
```
Comment (mcollina) [inline bloomrun.js]: if the bucket contains more than this element, this operation is wrong.

**Comment by mcollina** [comment]
Apart from the two little notes, it's all good.

**Comment by mcdonnelldean** [comment]
@mcollina super I'll get these worked out tonight and re commit

File: bloomrun.js (hunk 2 above)
Comment (mcdonnelldean) [inline bloomrun.js]: Do you mean the payload? It shouldn't matter to check it, it will be copyed if there. Or am I misunderstanding?

Comment (mcollina) [inline bloomrun.js]: Given the current API, you can add two elements with the same pattern, but different payload. `remove(payload)` should really be `remove(pattern, payload)`.

Comment (mcdonnelldean) [inline bloomrun.js]: Ah! Good catch, all add relevant tests too 😊

**Comment by mcdonnelldean** [comment]
@mcollina Ok I've updated, I had to break out some functions, could you let me know if you want them anywhere else? For now I just popped them at the top of the file. I added extra tests too.

File: README.md [hunk 4]
```diff
@@ -68,6 +69,15 @@ retrieval of objects. If no payload is provided the pattern itself will be
 returned.
 
 -------------------------------------------------------
+
+<a name="remove"></a>
+### instance.remove(pattern)
```
Comment (mcollina) [inline README.md]: Doc is missing the payload here

File: bloomrun.js [hunk 5]
```diff
@@ -43,6 +68,33 @@ BloomRun.prototype.add = function (pattern, payload) {
...
+      var bucket = matches[i]
+
+      if (removePattern(bucket, pattern, payload)) {
+        removeBucket(this._buckets, bucket)
+
+        var that = this
+
+        Object.keys(pattern).forEach(function (key) {
+          that._properties.delete(key)
+        })
```
Comment (mcollina) [inline bloomrun.js]: You can avoid setting `that` and creating a function here, declare them at top scope and pass a `this` parameter to `forEach`. See https://github.com/mcollina/bloomrun/blob/master/lib/genKeys.js for an example (and the docs!)

Allocating functions within loop is particularly bad!

**Pull request merged**